From 8f3d39c6f75767bbf4501427173a0e537809208c Mon Sep 17 00:00:00 2001
From: Valentin Catalin Neacsu <valentin-catalin.neacsu@nxp.com>
Date: Fri, 7 Sep 2018 10:48:45 +0300
Subject: [PATCH 678/767] net: phy: aquantia: Fix unknown speed reported by phy

Read the actual speed from PMA status register.

Signed-off-by: Valentin Catalin Neacsu <valentin-catalin.neacsu@nxp.com>
[Xulin: Original patch taken from NXP LSDK-18.12.]
Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 drivers/net/phy/aquantia.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/drivers/net/phy/aquantia.c b/drivers/net/phy/aquantia.c
index 96653e78bb71..39fd02b5c11a 100644
--- a/drivers/net/phy/aquantia.c
+++ b/drivers/net/phy/aquantia.c
@@ -232,7 +232,12 @@ static int aquantia_read_status(struct phy_device *phydev)
 	mdelay(10);
 	reg = phy_read_mmd(phydev, MDIO_MMD_PMAPMD, MDIO_CTRL1);
 
-	switch (reg &  MDIO_CTRL1_SPEEDSEL) {
+	if ((reg & MDIO_CTRL1_SPEEDSELEXT) == MDIO_CTRL1_SPEEDSELEXT)
+		reg &= MDIO_CTRL1_SPEEDSEL;
+	else
+		reg &= MDIO_CTRL1_SPEEDSELEXT;
+
+	switch (reg) {
 	case MDIO_PMA_CTRL1_AQ_SPEED5000:
 		phydev->speed = SPEED_5000;
 		break;
@@ -255,6 +260,7 @@ static int aquantia_read_status(struct phy_device *phydev)
 		phydev->speed = SPEED_UNKNOWN;
 		break;
 	}
+
 	phydev->duplex = DUPLEX_FULL;
 
 	return 0;
-- 
2.17.0

